web/client-api/src/main/java/io/deephaven/web/client/api/tree/JsTreeTable.java

```diff
@@ -209,455 +235,459 @@ public TreeRow(int offsetInSnapshot, Object[] dataColumns, Object rowStyleColumn
-                        child,
-                        depth + 1);
-                expandedMap.put(child, childState);
-                JsLog.debug("user expanded ", child);
-                scheduleSnapshotQuery(false);
-            }
-        }
-
-        public void collapse(Key child) {
-            if (expandedChildren.remove(child)) {
-                JsLog.debug("user collapsed ", child);
-
-                // from the entire tree's map, remove this child's expanded details,
-                // then iteratively remove each expanded child in the same way
-                List<TreeNodeState> removed = new ArrayList<>();
-                removed.add(expandedMap.remove(child));
-                while (!removed.isEmpty()) {
-                    TreeNodeState treeNode = removed.remove(0);
-                    for (Key expandedChild : treeNode.expandedChildren) {
-                        removed.add(expandedMap.remove(expandedChild));
-                    }
+            public void appendKeyData(Object[][] keyTableData, boolean expanded) {
+                int i;
+                for (i = 0; i < keyColumns.length; i++) {
+                    Js.<JsArray<Any>>cast(keyTableData[i]).push(keyColumns.getAt(i).get(this));
                 }
-
-                scheduleSnapshotQuery(false);
+                Js.<JsArray<Double>>cast(keyTableData[i++]).push((double) depth());
+                Js.<JsArray<Double>>cast(keyTableData[i++]).push(expanded ? ACTION_EXPAND : ACTION_COLLAPSE);
             }
         }
-
-        public TableDetails toTableDetails() {
-            TableDetails output = new TableDetails();
-            output.setKey(key);
-            output.setChildren(expandedChildren.toArray(new Key[0]));
-            return output;
-        }
     }
 
-    private final ClientTableState baseTable;
-    private WorkerConnection connection;
-
+    private final WorkerConnection connection;
+
+    // This group of fields represent the underlying state of the original HierarchicalTable
+    private final JsWidget widget;
+    private final InitialTableDefinition tableDefinition;
+    private final Column[] visibleColumns;
+    private final Map<String, Column> columnsByName = new HashMap<>();
+    private final int rowFormatColumn;
+    private final Map<String, Column> sourceColumns;
+    private final JsArray<Column> keyColumns = new JsArray<>();
+    private Column rowDepthCol;
+    private Column rowExpandedCol;
+    private final Column actionCol;
+    private final JsArray<Column> groupedColumns;
+
+    // The source JsTable behind the original HierarchicalTable, lazily built at this time
+    private final JsLazy<Promise<JsTable>> sourceTable;
+
+    // The current filter and sort state
     private List<FilterCondition> filters = new ArrayList<>();
     private List<Sort> sorts = new ArrayList<>();
+    private TicketAndPromise filteredTable;
+    private TicketAndPromise sortedTable;
+
+    private final Object[][] keyTableData;
```

Comment:
Could use a comment here about what this is for and why the size is 2 bigger than keyColumns, as it's not clear from just the name.